### PR TITLE
Regulate object links based on backend used

### LINKED ISSUE #125: Links in tickets still use /monitoring instead of /icingadb
We have migrated our DEV instance from IDO to IcingaDB some time ago.

Currently the JIRA tickets that are created by the module still have the /monitoring link in the description, even though the monitoring module isn't enabled anymore
image

Jira module version
1.3.3
e00284f0bf42b77b81854a1b198b65cba1a09eaa

OS. rhel8.8
PHP 7.2.24

In case further information is required don't hestitate asking.

best regards

### PR DESCRIPTION
**Pull request by raviks789**
fix #125

### CODE REVIEW

File: library/Jira/Web/Form/NewIssueForm.php [hunk 1]
```diff
@@ -275,7 +275,7 @@ protected function eventuallyAcknowledge($key, $host, $service)
         $ackMessage = "Jira issue $key has been created";
 
         try {
-            $cmd = new IcingaCommandPipe();
+            $cmd = (new IcingaCommandPipe($this->monitoringInfo));
```
Comment (nilmerg) [inline library/Jira/Web/Form/NewIssueForm.php:278]: here are no enclosing parentheses required at all

File: library/Jira/Web/RenderingHelper.php [hunk 2]
```diff
@@ -3,63 +3,179 @@
...
+     * Replace object urls in the given string with link elements
+     *
+     * @param string $string
+     *
+     * @return ?string
+     */
+    protected function replaceLinks($string)
+    {
+        return preg_replace_callback('/\[([^|]+)\|([^]]+)]/', function ($match) {
+            $url = Url::fromPath(htmlspecialchars_decode($match[2]));
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:124]: Doesn't `rawurldecode` suffice?

File: library/Jira/Web/RenderingHelper.php [hunk 3]
```diff
@@ -3,63 +3,179 @@
...
+            $link = (new Link(
+                $match[1],
+                $url,
+                ['target' => '_blank']
+            ));
+
+            if ($url->hasParam('service') || $url->hasParam('host.name')) {
+                if (str_contains($match[2], 'monitoring') && $this->useIcingaDbAsBackend()) {
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:132]: I have no idea why none of the tooling didn't complain about this, but `str_contains` is php 8 only.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Jira/Web/RenderingHelper.php (hunk 2 above)
Comment (raviks789) [reply library/Jira/Web/RenderingHelper.php:124]: this is because using `rawurldecode` did not help with decoding `&amp` in case there are multiple params.

Comment (raviks789) [reply library/Jira/Web/RenderingHelper.php]: As discussed link extraction and html injection have been separated.

File: library/Jira/Web/RenderingHelper.php [hunk 4]
```diff
@@ -3,63 +3,177 @@
...
+        $urls = [];
+
+        // Replace object urls in the given string with link elements
+        $body = preg_replace_callback('/\[([^|]+)\|([^]]+)]/', function ($match) use (&$urls) {
+            $url = Url::fromPath($match[2]);
+            $link = (new Link(
+                $match[1],
+                $url,
+                ['target' => '_blank']
+            ));
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:121]: Remove unnecessary parentheses!

File: library/Jira/Web/RenderingHelper.php [hunk 5]
```diff
@@ -3,63 +3,177 @@
...
+                }
+
+                $hostLink = clone $link;
+                $hostLink->setContent([$icon, $match[1]])
+                    ->addAttributes(['title' => t('Show Icinga Host State')]);
+                $this->setHostLink($hostLink);
+            }
+
+            $urls[] = $link->render();
+            return '$objectLink' . (count($urls) - 1) . '$';
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:166]: empty line before return

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Comment by nilmerg** [comment]
@sukhwinder33445 Please perform a test.

File: library/Jira/ProvidedHook/Icingadb/IcingadbSupport.php [hunk 6]
```diff
@@ -1,11 +0,0 @@
-<?php
-
-/* Icinga Web Jira Module | (c) 2021 Icinga GmbH | GPLv2*/
-
-namespace Icinga\Module\Jira\ProvidedHook\Icingadb;
-
-use Icinga\Module\Icingadb\Hook\IcingadbSupportHook;
-
-class IcingadbSupport extends IcingadbSupportHook
```
Comment (sukhwinder33445) [inline library/Jira/ProvidedHook/Icingadb/IcingadbSupport.php:9]: Please add again, this class is required to show the backend switch popup.

File: library/Jira/Common/IcingaDbSupport.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+// Icinga Web Jira Integration | (c) 2024 Icinga GmbH | GPLv2
+
+namespace Icinga\Module\Jira\Common;
+
+use Icinga\Application\Icinga;
+use Icinga\Application\Modules\Module;
+use Icinga\Authentication\Auth;
+
+trait IcingaDbSupport
```
Comment (sukhwinder33445) [inline library/Jira/Common/IcingaDbSupport.php:11]: Use the existing class instead, please. You can overwrite the method if required.

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

File: library/Jira/Common/IcingaDbSupport.php (hunk 7 above)
Comment (raviks789) [reply library/Jira/Common/IcingaDbSupport.php]: Using the existing class sounds good. There is no need to rewrite the method.

File: application/controllers/IssuesController.php [hunk 8]
```diff
@@ -67,7 +66,7 @@ protected function showNewIssueForm()
 
     protected function requireMonitoringInfo()
     {
-        if (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()) {
```
Comment (nilmerg) [inline application/controllers/IssuesController.php:70]: must not be removed

File: library/Jira/ProvidedHook/Icingadb/HostActions.php [hunk 9]
```diff
@@ -19,8 +19,8 @@ public function getActionsForObject(Host $host): array
                 Url::fromPath(
                     'jira/issues',
                     [
-                        'host'  => $host->name,
```
Comment (nilmerg) [inline library/Jira/ProvidedHook/Icingadb/HostActions.php:22]: restore formattting

File: library/Jira/ProvidedHook/Icingadb/ServiceActions.php [hunk 10]
```diff
@@ -19,9 +19,9 @@ public function getActionsForObject(Service $service): array
                 Url::fromPath(
                     'jira/issues',
                     [
-                        'service' => $service->name,
```
Comment (nilmerg) [inline library/Jira/ProvidedHook/Icingadb/ServiceActions.php:22]: restore formattting

File: library/Jira/ProvidedHook/Monitoring/HostActions.php [hunk 11]
```diff
@@ -19,8 +19,8 @@ public function getActionsForHost(Host $host)
             'Jira Issues' => Url::fromPath(
                 'jira/issues',
                 [
-                    'host' => $host->host_name,
```
Comment (nilmerg) [inline library/Jira/ProvidedHook/Monitoring/HostActions.php:22]: restore formattting

File: library/Jira/ProvidedHook/Monitoring/ServiceActions.php [hunk 12]
```diff
@@ -19,9 +19,9 @@ public function getActionsForService(Service $service)
             'Jira Issues' => Url::fromPath(
                 'jira/issues',
                 [
-                    'host'    => $service->host_name,
```
Comment (nilmerg) [inline library/Jira/ProvidedHook/Monitoring/ServiceActions.php:22]: restore formattting

File: library/Jira/Web/RenderingHelper.php [hunk 13]
```diff
@@ -17,49 +18,161 @@ class RenderingHelper
...
+        $body = preg_replace_callback('/\[([^|]+)\|([^]]+)]/', function ($match) use (&$urls) {
+            $url = Url::fromPath($match[2]);
+            $link = new Link(
+                $match[1],
+                $url,
+                ['target' => '_blank']
+            );
+
+            if ($url->hasParam('service') || $url->hasParam('host.name')) {
+                if (strpos($match[2], 'icingaweb2/monitoring') !== false && IcingadbSupport::useIcingaDbAsBackend()) {
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:122]: only if icingadb is enabled

File: library/Jira/Web/RenderingHelper.php [hunk 14]
```diff
@@ -17,49 +18,161 @@ class RenderingHelper
...
+                }
+
+                $serviceLink = clone $link;
+                $serviceLink->setContent([new Icon('cog'), $match[1]])
+                    ->addAttributes(['title' => t('Show Icinga Service State')]);
+                $this->setServiceLink($serviceLink);
+            } else {
+                $icon = new Icon('server');
+                if (strpos($match[2], 'icingaweb2/monitoring') !== false) {
+                    if (IcingadbSupport::useIcingaDbAsBackend()) {
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:143]: only if icingadb is enabled

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]
Seems fine. But please make another test @sukhwinder33445

**Review by sukhwinder33445 — COMMENTED** [review]
Are the following points intentional or bugs?
* Link to host/service in the issues list is opened in a new tab instead of the second column.
* Issues created with icingadb will not change the link if icingadb is disabled.

**Comment by raviks789** [comment]
> Are the following points intentional or bugs?
> 
> * Link to host/service in the issues list is opened in a new tab instead of the second column.
> * Issues created with icingadb will not change the link if icingadb is disabled.

>  Link to host/service in the issues list is opened in a new tab instead of the second column -> this should be fixed, only external links should be opened in new tab
>  Issues created with icingadb will not change the link if icingadb is disabled -> this is intentional

**Review by sukhwinder33445 — COMMENTED** [review]
Manually adding some links to issue description leads to wrong host/service value/link.
Screenshot 2024-04-12 at 16 15 18

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

File: library/Jira/Web/RenderingHelper.php [hunk 15]
```diff
@@ -17,49 +19,211 @@ class RenderingHelper
...
+            $html = str_replace('$objectLink' . $i . '$', $url, $html);
         }
 
-        return new Link([new Icon('cog'), $service], Url::fromPath('monitoring/service/show', [
-            'host'    => $host,
-            'service' => $service,
-        ]), [
-            'title' => t('Show Icinga Service State'),
-        ]);
+        // This is safe.
```
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Jira/Web/RenderingHelper.php:225]: This is something from the old code. I think it means returning the string as `HtmlString`.

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Jira/Web/RenderingHelper.php:225]: please remove, it is not really helpful.

File: library/Jira/Web/Table/IssueDetails.php [hunk 16]
```diff
@@ -41,10 +40,12 @@ protected function assemble()
         $icingaKey = preg_replace('/^BEGIN(.+)END$/', '$1', $fields->$keyField);
         $parts = explode('!', $icingaKey);
         $host = array_shift($parts);
+        $this->helper->setHostName($host);
```
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Jira/Web/Table/IssueDetails.php]: I understand what you mean, but this results in making more changes than needed.

**Review by sukhwinder33445 — DISMISSED** [review]
Works fine.

File: library/Jira/Web/RenderingHelper.php [hunk 17]
```diff
@@ -17,49 +19,207 @@ class RenderingHelper
...
+                        $urlHostParam = $url->getParam('name');
+                    }
+
+                    if ($urlHostParam === $this->hostName) {
+                        $hostLink = clone $link;
+                        $hostLink->setContent([$icon, $match[1]])
+                            ->addAttributes(['title' => t('Show Icinga Host State')]);
+                        $this->setHostLink($hostLink);
+                    }
+                }
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:206]: This has still numerous problems:

* `icingaweb2/` is **your** base URL, but that isn't always the case
  * Use `$url->getPath()` instead of `$match[2]`, it only returns the relative path
* Looking for `service` anywhere in the URL may lead to false positives
  * Use either `$url->getPath()` to search for `/service` or `$url->hasParam('service')` for a parameter
  * But don't transform `monitoring/host/show?host=docker-master&service=disk` to an Icinga DB Web service URL just because there's a `service` param
* This transforms any monitoring link now, not just the one that got automatically added. So there might be a chance that additional parameters are part of the URL. Overriding the link's URL with a completely new URL drops those.
  * If you transform from monitoring to Icinga DB Web, specifically override the path with `$url->setPath()`, remove old parameters with `$url->remove()` and add new ones with `$url->setParams()`

File: library/Jira/Web/RenderingHelper.php [hunk 18]
```diff
@@ -17,49 +19,207 @@ class RenderingHelper
...
+            } elseif ($url->isExternal()) {
+                $link->addAttributes(['target' => '_blank']);
+            }
+
+            $urls[] = $link->render();
+
+            return '$objectLink' . (count($urls) - 1) . '$';
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:213]: Is there a reason why this doesn't return the transformed URL already?

File: library/Jira/Web/RenderingHelper.php [hunk 19]
```diff
@@ -17,49 +19,207 @@ class RenderingHelper
...
+        }, $body);
+
+        $html = Html::wantHtml($body)->render();
```
Comment (nilmerg) [inline library/Jira/Web/RenderingHelper.php:216]: `$body` is already a string, why is this needed?

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Jira/Web/RenderingHelper.php (hunk 18 above)
Comment (raviks789) [reply library/Jira/Web/RenderingHelper.php]: Yes, as you said the transformed url can be directly returned.

File: library/Jira/Web/RenderingHelper.php (hunk 19 above)
Comment (raviks789) [reply library/Jira/Web/RenderingHelper.php]: Not needed right now.

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

File: library/Jira/Web/RenderingHelper.php [hunk 20]
```diff
@@ -17,49 +19,200 @@ class RenderingHelper
...
+                    && Module::exists('icingadb')
+                    && IcingadbSupport::useIcingaDbAsBackend();
+                if (strpos($urlPath, '/service') !== false) {
+                    if ($monitoringObjectLink) {
+                        $urlServiceParam = $url->getParam('service');
+                        $urlHostParam = $url->getParam('host');
+                        if ($transformToIcingadbLink) {
+                            $url->setPath('icingadb/service')
+                                ->remove(['service', 'host'])
+                                ->setParams(['name' => $urlServiceParam, 'host.name' => $urlHostParam]);
```
Comment (sukhwinder33445) [inline library/Jira/Web/RenderingHelper.php:169]: Please use `addParams`.

File: library/Jira/Web/RenderingHelper.php [hunk 21]
```diff
@@ -17,49 +19,200 @@ class RenderingHelper
...
+                } else {
+                    $icon = new Icon('server');
+                    if ($monitoringObjectLink) {
+                        $urlHostParam = $url->getParam('host');
+                        if ($transformToIcingadbLink) {
+                            $url->setPath('icingadb/host')
+                                ->remove('host')
+                                ->setParams(['name' => $urlHostParam]);
+
+
```
Comment (sukhwinder33445) [inline library/Jira/Web/RenderingHelper.php:192]: Please remove this extra empty new line.

File: library/Jira/Web/RenderingHelper.php [hunk 22]
```diff
@@ -17,49 +19,200 @@ class RenderingHelper
...
+                        $this->setServiceLink($serviceLink);
+                    }
+                } else {
+                    $icon = new Icon('server');
+                    if ($monitoringObjectLink) {
+                        $urlHostParam = $url->getParam('host');
+                        if ($transformToIcingadbLink) {
+                            $url->setPath('icingadb/host')
+                                ->remove('host')
+                                ->setParams(['name' => $urlHostParam]);
```
Comment (sukhwinder33445) [inline library/Jira/Web/RenderingHelper.php:190]: Please use `addParams`.

File: library/Jira/Web/RenderingHelper.php [hunk 23]
```diff
@@ -17,49 +19,200 @@ class RenderingHelper
...
+    public function formatBody(string $body): HtmlString
+    {
+        // Replace object urls in the given string with link elements
+        $body = preg_replace_callback('/\[([^|]+)\|([^]]+)]/', function ($match) {
+            $url = Url::fromPath($match[2]);
+            $link = new Link($match[1], $url);
+            $urlPath = $url->getPath();
+
+            $monitoringObjectLink = in_array($urlPath, ['monitoring/host/show', 'monitoring/service/show']);
+
```
Comment (sukhwinder33445) [inline library/Jira/Web/RenderingHelper.php:153]: Please remove this new line.

File: library/Jira/Web/RenderingHelper.php [hunk 24]
```diff
@@ -17,49 +19,200 @@ class RenderingHelper
...
+            Html::tag('h3', null, Html::sprintf(
+                '%s: %s',
+                $this->shortTimeSince($time),
+                $author
+            )),
+            Html::tag(
+                'pre',
+                ['class' => 'comment'],
+                $this->formatBody($body)
+            ),
```
Comment (sukhwinder33445) [inline library/Jira/Web/RenderingHelper.php:133]: This can be done in one line.

**Review by sukhwinder33445 — APPROVED** [review]

**Review by nilmerg — APPROVED** [review]
